### PR TITLE
Adding tests for the Source Control cmdlets and enabling Update Management tests

### PR DESCRIPTION
**Pull request by Francisco-Gamino**
## Description
* Adding tests for the Source Control cmdlets
* Enabling tests for the Update Management cmdlets

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by Francisco-Gamino** [comment]
Adding @finiteattractor (owner for Update Management) for FYI.

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+# Here is the sequence:
+# 1) Create source control - validate properties
+# 2) Update the newly created source control - validate that is was updated
+# 3) Start a sync - validate that the sync completes 
+# 4) Retrieve the streams for the completed source control sync job
+# 5) Clean up - remove the source control
+
+# Automation account information
+$resourceGroupName = "frangom-test"
+$subscriptionId = "52d8cf1b-bcac-493a-bbae-f234b5ff38b0"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: why is subscriptionId hard-coded in the tests?  This is an anti-pattern, it makes it impossible for contributors to run the tests

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        [ValidateNotNullOrEmpty()]
+        [String]
+        $ExpectedState
+    )
+
+    $state = ""
+    $waitTimeInSeconds = 5
+    $retries = 20
+
+    do
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: There is a common code Retry method that does precisely this

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+#endregion
+
+
+function Test-CreateVsoGitSourceControlAndSync
+{
+    # VsoGit repo info
+    $sourceControl = @{
+        Name = "AASourceControl-VsoGit"
+        RepoUrl = "https://francisco-gamino.visualstudio.com/_git/VsoGit-SwaggerAndCmdletsTests"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: ?? This should either be configurable, or a public repo available to everyone

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        FolderPath = "Azure/MyRunbooks"
+        SourceType = "VsoGit"
+    }
+
+    try
+    {
+        EnsureSourceControlDoesNotExist -Name $sourceControl.Name
+
+        # Access token
+        $token = "3qdxa22lutnhezd4atpna74jn3m7wgo6o6kfbwezjfnvgbjhvoca"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: Is this a credential?

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        $updatedSourceControl = Update-AzureRmAutomationSourceControl -ResourceGroupName $resourceGroupName `
+                                                                      -AutomationAccountName $automationAccountName  `
+                                                                      -Name $sourceControl.Name  `
+                                                                      -PublishRunbook $true
+        $expectedPropertyValue = "True"
+        Assert-AreEqual $updatedSourceControl.PublishRunbook $expectedPropertyValue `
+                "'PublishRunbook' property does not match. Expected: $expectedPropertyValue. Actual: $($updatedSourceControl.PublishRunbook)"
+        
+        # Start a sync for the source control
+        $jobId = "0bfa6b49-c08c-4b2f-853e-08128c3c86ee"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: This should not be hardcoded - you should be able to query for the job id

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        EnsureSourceControlDoesNotExist -Name $sourceControl.Name
+    }
+}
+
+function Test-CreateVsoTfvcSourceControlAndSync
+{
+    # VsoTfvc source control info
+    $sourceControl = @{
+        Name = "AASourceControl-VsoTfvc"
+        RepoUrl = "https://francisco-gamino.visualstudio.com/VsoTfvc-SwaggerAndCmdletsTests/_versionControl"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: see above

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+        FolderPath = "/MyRunbooks"
+        SourceType = "VsoTfvc"
+    }
+
+    try
+    {
+        EnsureSourceControlDoesNotExist -Name $sourceControl.Name
+
+        # Access token
+        $token = "3qdxa22lutnhezd4atpna74jn3m7wgo6o6kfbwezjfnvgbjhvoca"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: see above

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        EnsureSourceControlDoesNotExist -Name $sourceControl.Name
+    }
+}
+
+function Test-CreateGitHubSourceControlAndSync
+{
+    # GitHub repo info
+    $sourceControl = @{
+        Name = "AASourceControl-GitHub"
+        RepoUrl = "https://github.com/Francisco-Gamino/SwaggerAndCmdletsTests.git"
```
Comment (markcowl) [inline src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: see above - assume that the comments aboud hard coded ids, repos, etc. apply throughout the tests

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 (hunk 2 above)
Comment (Francisco-Gamino) [reply src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: Thanks @markcowl, could you please point me to the code?

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 (hunk 3 above)
Comment (Francisco-Gamino) [reply src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: Automation Source Control links your Automation account with their GitHub or VSTS (Azure Repos). This feature only works with Private repos.

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 (hunk 4 above)
Comment (Francisco-Gamino) [reply src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: This is a Personal Access Token which is required to access GitHub or VSTS.

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 (hunk 5 above)
Comment (Francisco-Gamino) [reply src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: This job id is part of the Url for the call. Originally, I was creating one on the fly (New-Guid).Guid, but this was causing my reply to fail.

File: src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1 (hunk 1 above)
Comment (Francisco-Gamino) [reply src/ResourceManager/Automation/Commands.Automation.Test/ScenarioTests/SourceControlTests.ps1]: Thanks @markcowl for pointing this out. My test requires a resource group, a subscription id, and a automation account. What is the best way to provide this info for the tests. Could you please point me to an example?

**Comment by MiYanni** [comment]
@Francisco-Gamino It looks like the 2 tests are failing (likely) because of a date encoding issue. It is probably trying to match for a 4 digit year when only a 2 digit year is written (or visa versa).

**Comment by Francisco-Gamino** [comment]
> @Francisco-Gamino It looks like the 2 tests are failing (likely) because of a date encoding issue. It is probably trying to match for a 4 digit year when only a 2 digit year is written (or visa versa).

Thanks @MiYanni, I will follow up with @finiteattractor to get this fixed.

**Comment by Francisco-Gamino** [comment]
@MiYanni : I did a bit more of research opened issue https://github.com/Azure/azure-powershell/issues/7705. After this issue is fixed, @finiteattractor will have to re-record the tests. For now, I am skipping them in netstandard. 

@markcowl : Could you please take another look? 

Note: I need to get this PR merge today if possible. I am transitioning to a new team in the next couple of days. Thanks!

**Comment by MiYanni** [comment]
@azuresdkci Retest this please

**Comment by MiYanni** [comment]
@Francisco-Gamino It seems that the same tests failed on net452 (Windows) too. Not sure what the issue is, but it isn't finding the entry properly.

**Comment by MiYanni** [comment]
@azuresdkci Retest this please

**Comment by MiYanni** [comment]
@azuresdkci Retest this please

**Comment by Francisco-Gamino** [comment]
The build finally passed!
@markcowl , @MiYanni : Could you please take another look? Thanks!

**Pull request merged**